libopenage/gamestate/event/drag_select.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023-2023 the openage authors. See copying.md for legal info.
+
+#include "drag_select.h"
+
+#include <eigen3/Eigen/Dense>
+
+#include "coord/phys.h"
+#include "coord/pixel.h"
+#include "coord/scene.h"
+#include "curve/discrete.h"
+#include "gamestate/component/internal/ownership.h"
+#include "gamestate/component/internal/position.h"
+#include "gamestate/game_entity.h"
+#include "gamestate/game_state.h"
+#include "gamestate/types.h"
+
+
+namespace openage::gamestate::event {
+
+DragSelectHandler::DragSelectHandler() :
+	OnceEventHandler{"game.drag_select"} {}
+
+void DragSelectHandler::setup_event(const std::shared_ptr<openage::event::Event> & /* event */,
+                                    const std::shared_ptr<openage::event::State> & /* state */) {
+	// TODO
+}
+
+void DragSelectHandler::invoke(openage::event::EventLoop & /* loop */,
+                               const std::shared_ptr<openage::event::EventEntity> & /* target */,
+                               const std::shared_ptr<openage::event::State> &state,
+                               const time::time_t &time,
+                               const param_map &params) {
+	auto gstate = std::dynamic_pointer_cast<openage::gamestate::GameState>(state);
+
+	size_t controlled_id = params.get("controlled", 0);
+
+	Eigen::Matrix4f id_matrix = Eigen::Matrix4f::Identity();
+	Eigen::Matrix4f cam_matrix = params.get("camera_matrix", id_matrix);
+	Eigen::Vector2f drag_start = params.get("drag_start", Eigen::Vector2f{0, 0});
+	Eigen::Vector2f drag_end = params.get("drag_end", Eigen::Vector2f{0, 0});
+
+	// Boundaries of the rectangle
+	float top = std::max(drag_start.y(), drag_end.y());
+	float bottom = std::min(drag_start.y(), drag_end.y());
+	float left = std::min(drag_start.x(), drag_end.x());
+	float right = std::max(drag_start.x(), drag_end.x());
+
+	log::log(DBG << "Drag select rectangle (NDC):");
+	log::log(DBG << "\tTop: " << top);
+	log::log(DBG << "\tBottom: " << bottom);
+	log::log(DBG << "\tLeft: " << left);
+	log::log(DBG << "\tRight: " << right);
+
+	std::vector<entity_id_t> selected;
+	for (auto &entity : gstate->get_game_entities()) {
```

Comment:
for now ok, but we definitely need an index structure from position -> entity.

Reply:
true that